Foundation/NSPathUtilities.swift

```diff
@@ -446,57 +446,108 @@ public extension NSString {
+        let checkExtension = _getExtensionChecker(filterTypes, caseSensetive: flag)
         
-        guard let urlWhereToSearch = searchAllFilesInDirectory ? url : url.URLByDeletingLastPathComponent else {
+        guard let
+            resolvedURL = url._resolveSymlinksInPath(excludeSystemDirs: false),
+            urlWhereToSearch = searchAllFilesInDirectory ? resolvedURL : resolvedURL.URLByDeletingLastPathComponent
+        else {
             return 0
         }
-        
-        var matches: [String] = []
-        
-        let namePrefix = url.lastPathComponent ?? ""
 
-        let enumerator = NSFileManager.defaultManager().enumeratorAtURL(urlWhereToSearch, includingPropertiesForKeys: nil, options: .SkipsSubdirectoryDescendants, errorHandler: nil)
+        var matches = _getNamesAtURL(urlWhereToSearch, prependWith: "", nameValidator: checkFileName, typeValidator: checkExtension)
         
-        while let item = enumerator?.nextObject() as? NSURL {
-            
-            let itemName = item.lastPathComponent ?? ""
-            let itemExtension = item.pathExtension ?? ""
-            let normalizedExtension = flag ? itemExtension : itemExtension.lowercaseString
-            
-            let matchByName = searchAllFilesInDirectory || itemName.bridge().rangeOfString(namePrefix, options: compareOptions).location == 0
-            let matchByExtension = types.isEmpty || types.contains(normalizedExtension)
-            
-            if matchByName && matchByExtension {
-                matches.append(itemName)
+        if matches.count == 1 {
+            let theOnlyFoundItem = NSURL(fileURLWithPath: matches[0], relativeToURL: urlWhereToSearch)
+            if theOnlyFoundItem.hasDirectoryPath {
+                matches = _getNamesAtURL(theOnlyFoundItem, prependWith: matches[0], nameValidator: { _ in true }, typeValidator: checkExtension)
             }
         }
         
-        let commonPath = urlWhereToSearch.absoluteString!.bridge().stringByReplacingOccurrencesOfString("file://", withString: "")
+        let commonPath = searchAllFilesInDirectory ? _storage : _ensureLastPathSeparator(stringByDeletingLastPathComponent)
         
-        if let lcp = _longestCommonPrefix(matches, caseSensitive: flag) {
-           outputName = (commonPath + lcp).bridge()
+        if searchAllFilesInDirectory {
+            outputName = "/"
+        } else {            
+            if let lcp = _longestCommonPrefix(matches, caseSensitive: flag) {
+                outputName = (commonPath + lcp).bridge()
+            }
         }
         
         outputArray = matches.map({ (commonPath + $0).bridge() })
         
         return matches.count
     }
+
+    internal func _stringIsPathToDirectory(path: String) -> Bool {
+        if !path.hasSuffix("/") {
+            return false
+        }
+        
+        var isDirectory = false
+        let isAbsolutePath = NSFileManager.defaultManager().fileExistsAtPath(path, isDirectory: &isDirectory)
```

Comment:
`isAbsolutePath` seems like a misleading name; `exists` perhaps just to avoid future confusion?